sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,335 @@
+    : Double = {
+
+    val aColStat = planStat.colStats(attrRef.name)
+    val ndv = aColStat.distinctCount
+
+    /**
+     * decide if the value is in [min, max] of the column.
+     * We currently don't store min/max for binary/string type.
+     * Hence, we assume it is in boundary for binary/string type.
+     */
+    val inBoundary: Boolean = attrRef.dataType match {
+      case dataType: DataType if (dataType.isInstanceOf[NumericType]
+        || dataType.isInstanceOf[DateType]
+        || dataType.isInstanceOf[TimestampType]) =>
+        val statsRange = Range(aColStat.min, aColStat.max, dataType).asInstanceOf[NumericRange]
+
+        attrRef.dataType match {
+          case intType: DataType if intType.isInstanceOf[IntegralType] =>
+            (BigDecimal(literal.value.asInstanceOf[Long]) >= statsRange.min) &&
+              (BigDecimal(literal.value.asInstanceOf[Long]) <= statsRange.max)
+          case fracType: DataType if fracType.isInstanceOf[FractionalType] =>
+            (BigDecimal(literal.value.asInstanceOf[Double]) >= statsRange.min) &&
+              (BigDecimal(literal.value.asInstanceOf[Double]) <= statsRange.max)
+          case DateType =>
+            val dateLiteral = DateTimeUtils.stringToDate(literal.value.asInstanceOf[UTF8String])
+            if (dateLiteral.isEmpty) {
+              logDebug("[CBO] Date literal is wrong, No statistics for " + attrRef)
+              return 1.0
+            }
+            val dateBigDecimal = BigDecimal(dateLiteral.asInstanceOf[BigInt])
+            (dateBigDecimal >= statsRange.min) && (dateBigDecimal <= statsRange.max)
+          case TimestampType =>
+            val tsLiteral = DateTimeUtils.stringToTimestamp(literal.value.asInstanceOf[UTF8String])
+            if (tsLiteral.isEmpty) {
+              logDebug("[CBO] Timestamp literal is wrong, No statistics for " + attrRef)
+              return 1.0
+            }
+            val tsBigDecimal = BigDecimal(tsLiteral.asInstanceOf[BigInt])
+            (tsBigDecimal >= statsRange.min) && (tsBigDecimal <= statsRange.max)
+        }
+
+      case _ => true  /** for String/Binary type */
+    }
+
+    val percent: Double =
+      if (inBoundary) {
+
+        if (update) {
+          /**
+           * We update ColumnStat structure after apply this equality predicate.
+           * Set distinctCount to 1.  Set nullCount to 0.
+           */
+          val newStats = attrRef.dataType match {
+            case dataType: DataType if (dataType.isInstanceOf[NumericType]
+              || dataType.isInstanceOf[DateType]
+              || dataType.isInstanceOf[TimestampType]) =>
+              val newValue = Some(literal.value)
+              aColStat.copy(distinctCount = 1, min = newValue,
+                max = newValue, nullCount = 0)
+            case _ => aColStat.copy(distinctCount = 1, nullCount = 0)
+          }
+          mutableColStats += (attrRef.name -> newStats)
+        }
+
+        1.0 / ndv.toDouble

```

Comment:
Seems it's the only way to estimate percent with basic column stats, though this could under-estimate dramatically.